ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala

```diff
@@ -0,0 +1,149 @@
+import com.daml.ledger.api.v1.admin.participant_pruning_service.{
+  ParticipantPruningServiceGrpc,
+  PruneRequest
+}
+import com.daml.ledger.api.v1.ledger_offset.LedgerOffset
+import com.daml.ledger.participant.state.index.v2.{IndexParticipantPruningService, LedgerEndService}
+import com.daml.ledger.participant.state.v1.{
+  NotPruned,
+  Offset,
+  ParticipantPruned,
+  SubmissionId,
+  WriteParticipantPruningService
+}
+import com.daml.lf.data.Ref
+import com.daml.logging.{ContextualizedLogger, LoggingContext}
+import com.daml.platform.api.grpc.GrpcApiService
+import com.daml.platform.server.api.validation.ErrorFactories
+import com.google.protobuf.empty.Empty
+import io.grpc.{ServerServiceDefinition, StatusRuntimeException}
+
+import scala.compat.java8.FutureConverters
+import scala.concurrent.{ExecutionContext, Future}
+
+final class ApiParticipantPruningService private (
+    readBackend: IndexParticipantPruningService with LedgerEndService,
+    writeBackend: WriteParticipantPruningService)(
+    implicit grpcExecutionContext: ExecutionContext,
+    logCtx: LoggingContext)
+    extends ParticipantPruningServiceGrpc.ParticipantPruningService
+    with GrpcApiService {
+
+  private val logger = ContextualizedLogger.get(this.getClass)
+
+  override def bindService(): ServerServiceDefinition =
+    ParticipantPruningServiceGrpc.bindService(this, DirectExecutionContext)
+
+  override def prune(request: PruneRequest): Future[Empty] = {
+    val submissionIdOrErr = SubmissionId
+      .fromString(
+        if (request.submissionId.nonEmpty) request.submissionId else UUID.randomUUID().toString)
+      .left
+      .map(err => ErrorFactories.invalidArgument(s"submission_id $err"))
+
+    submissionIdOrErr.fold(
+      Future.failed,
+      submissionId =>
+        LoggingContext.withEnrichedLoggingContext("submissionId" -> submissionId) {
+          implicit logCtx =>
+            (for {
+
+              pruneUpTo <- validateRequest(request: PruneRequest)
+
+              _ <- pruneWriteService(pruneUpTo, submissionId)
+
+              pruneResponse <- pruneLedgerApiServerIndex(pruneUpTo)
+
+            } yield pruneResponse).andThen(logger.logErrorsOnCall[Empty])
+      }
+    )
+  }
+
+  private def validateRequest(request: PruneRequest)(
+      implicit logCtx: LoggingContext): Future[Offset] = {
+    (for {
+      pruneUpToProto <- checkOffsetIsSpecified(request.pruneUpTo)
+      pruneUpToString <- checkOffsetIsAbsolute(pruneUpToProto)
+      pruneUpTo <- checkOffsetIsHexadecimal(pruneUpToString)
+    } yield (pruneUpTo, pruneUpToString))
+      .fold(Future.failed, o => checkOffsetIsBeforeLedgerEnd(o._1, o._2))
+  }
+
+  private def pruneWriteService(pruneUpTo: Offset, submissionId: SubmissionId)(
+      implicit logCtx: LoggingContext): Future[Unit] =
+    FutureConverters
+      .toScala(writeBackend.prune(pruneUpTo, submissionId))
+      .flatMap {
+        case NotPruned(status) => Future.failed(ErrorFactories.grpcError(status))
```

Comment:
As a failed admin operation, it should probably print a log line at the highest log level.

Reply:
Agreed with logging an error in these cases. This should be taken care of by this expression following the future-for-comprehension in the caller or?: `.andThen(logger.logErrorsOnCall[Empty])`

Reply:
Ah right, I missed that part. There is one small detail, `logger.logErrorsOnCall` may not contain details on which part of the pruning failed (e.g., whether it was the index pruning or the write service pruning), but I trust WriteService implementations to put enough details in the Status message.